src/controller/tests/data_model/TestRead.cpp

```diff
@@ -295,6 +354,316 @@ void TestReadInteraction::TestReadDataVersionFilter(nlTestSuite * apSuite, void
+    attributePathParams2[1].mClusterId   = Test::MockClusterId(2);
+    attributePathParams2[1].mAttributeId = Test::MockAttributeId(2);
+
+    chip::app::AttributePathParams attributePathParams3[3];
+    attributePathParams3[0].mEndpointId  = Test::kMockEndpoint1;
+    attributePathParams3[0].mClusterId   = Test::MockClusterId(2);
+    attributePathParams3[0].mAttributeId = kInvalidAttributeId;
+
+    attributePathParams3[1].mEndpointId  = Test::kMockEndpoint2;
+    attributePathParams3[1].mClusterId   = Test::MockClusterId(3);
+    attributePathParams3[1].mAttributeId = kInvalidAttributeId;
+
+    attributePathParams3[2].mEndpointId  = Test::kMockEndpoint2;
+    attributePathParams3[2].mClusterId   = Test::MockClusterId(2);
+    attributePathParams3[2].mAttributeId = kInvalidAttributeId;
+
+    chip::app::EventPathParams eventPathParams[100];
+    for (uint32_t index = 0; index < 100; index++)
+    {
+        eventPathParams[index].mEndpointId = Test::kMockEndpoint3;
+        eventPathParams[index].mClusterId  = Test::MockClusterId(2);
+        eventPathParams[index].mEventId    = 0;
+    }
+
+    chip::app::ReadPrepareParams readPrepareParams(ctx.GetSessionBobToAlice());
+    readPrepareParams.mMinIntervalFloorSeconds   = 0;
+    readPrepareParams.mMaxIntervalCeilingSeconds = 4;
+    //
+    // Test the application callback as well to ensure we get the right number of SubscriptionEstablishment/Termination
+    // callbacks.
+    //
+    app::InteractionModelEngine::GetInstance()->RegisterReadHandlerAppCallback(&gTestReadInteraction);
+
+    int testId = 0;
+    // Initial Read of C1A1, C1A2 and C2A1. Cache has two C1 attributes at version 0, and one C2 attribute at version 0.
+    // Expect no versions would be cached.
+    {
+        testId++;
+        ChipLogProgress(DataManagement, "\t -- Running Read with AttributeCache Test ID %d", testId);
+        app::ReadClient readClient(chip::app::InteractionModelEngine::GetInstance(), &ctx.GetExchangeManager(),
+                                   cache.GetBufferedCallback(), chip::app::ReadClient::InteractionType::Read);
+        readPrepareParams.mpAttributePathParamsList    = attributePathParams1;
+        readPrepareParams.mAttributePathParamsListSize = 3;
+        err                                            = readClient.SendRequest(readPrepareParams);
+        NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+        ctx.DrainAndServiceIO();
+        NL_TEST_ASSERT(apSuite, delegate.mNumAttributeResponse == 3);
+        NL_TEST_ASSERT(apSuite, !delegate.mReadError);
+        Optional<DataVersion> version;
+        NL_TEST_ASSERT(apSuite, cache.GetVersion(Test::kMockEndpoint2, Test::MockClusterId(3), version) == CHIP_NO_ERROR);
+        NL_TEST_ASSERT(apSuite, !version.HasValue());
+        delegate.mNumAttributeResponse = 0;
+    }
+    // Second read of C1A* and C2A1. We cannot use the stored data versions in the cache since our subsequent read is 'broader' then
```

Comment:
We have no stored data versions in the cache right?

Reply:
right, update the description